File: Assignment 2/Python/2_question.py
"""
2. Write a program to sort a list of element in an Array using Merge Sort. Show the
running time complexity w.r.t different input cases (best/average/worst).
"""

from timeit import default_timer as timer


def merge_sort(array):
    if len(array) > 1:
        r = len(array)//2
        L = array[:r]
        M = array[r:]
        merge_sort(L)
        merge_sort(M)
        i = j = k = 0
        while i < len(L) and j < len(M):
            if L[i] < M[j]:
                array[k] = L[i]
                i += 1
            else:
                array[k] = M[j]
                j += 1
            k += 1

        while i < len(L):
            array[k] = L[i]
            i += 1
            k += 1

        while j < len(M):
            array[k] = M[j]
            j += 1
            k += 1


def print_list(array):
    for i in range(len(array)):
        print(array[i], end=" ")
    print()



array = [6, 5, 12, 10, 9, 1]

start = timer()
merge_sort(array)
end = timer()

print("Sorted array is: ")
print_list(array)
print(end - start)

File: Assignment 2/C/2_question.c
/*
2. Write a program to sort a list of element in an Array using Merge Sort. Show the
running time complexity w.r.t different input cases (best/average/worst).
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void merge(int arr[], int l, int m, int r) {
    int i, j, k;
    int n1 = m - l + 1;
    int n2 = r - m;
    int L[n1], R[n2];
 
    for (i = 0; i < n1; ++i)
        L[i] = arr[l + i];

    for (j = 0; j < n2; ++j)
        R[j] = arr[m + 1 + j];
 
    i = 0;
    j = 0;
    k = l;
    while (i < n1 && j < n2)
    {
        if (L[i] <= R[j])
        {
            arr[k] = L[i];
            ++i;
        }

        else
        {
            arr[k] = R[j];
            ++j;
        }

        ++k;
    }

    while (i < n1)
    {
        arr[k] = L[i];
        ++i;
        ++k;
    }

    while (j < n2)
    {
        arr[k] = R[j];
        ++j;
        ++k;
    }
}

void mergeSort(int arr[], int l, int r) {
    if (l < r)
    {
        int m = l + (r - l) / 2;

        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

void printArray(int A[], int size)
{
    int i;
    for (i = 0; i < size; ++i)
        printf("%d ", A[i]);
    
    printf("\n");
}

int main() {
    int arr[] = { 12, 11, 13, 5, 6, 7 };
    int arr_size = sizeof(arr) / sizeof(arr[0]);
    double time_spent;
    clock_t start, end;
 
    printf("Given array is \n");
    printArray(arr, arr_size);

    start = clock();
    mergeSort(arr, 0, arr_size - 1);
    end = clock();

    printf("\nSorted array is \n");
    printArray(arr, arr_size);

    time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    printf("\n The elapsed time is %f seconds", time_spent);

    return 0;
}